### PR TITLE
WebSocket layer to be able to connect "web" client

### PR DESCRIPTION
**Pull request by stumpyfr**
Currently working with a mini irc js implem, the flow:
- PASS
- NICK
- USER
- JOIN
- PRIVMSG

works and the ping/pong timeout keep the communication open.

### CODE REVIEW

**Comment by edmund-huber** [comment]
Nice. I will test this later tonight. Is your frontend also opensource? Can you give / show me your set up so I can test it the way you tested it?

**Comment by stumpyfr** [comment]
Sadly not...no yet. I will open the irc js part, not all the client.
To test, I used on my side: https://github.com/gorilla/websocket/blob/master/examples/chat/home.html

Need to send all the commands and after the USER, you will get the reply with the MOTD and all the "welcome" payload.

Not the easiest way I need to admit but working :)

File: ergonomadic.conf [hunk 1]
```diff
@@ -1,8 +1,9 @@
 [server]
-name = "irc.example.com" ; required, usually a hostname
```
Comment (edmund-huber) [inline ergonomadic.conf:2]: This change wasn't necessary. Undo?

File: irc/server.go [hunk 2]
```diff
@@ -72,6 +73,10 @@ func NewServer(config *Config) *Server {
 		server.listen(addr)
 	}
 
+	if config.Server.Wslisten != "" {
```
Comment (edmund-huber) [inline irc/server.go:76]: For consistency, do you think we should allow multiple `wslisten` like we do for `listen`?

File: irc/server.go [hunk 3]
```diff
@@ -204,6 +209,37 @@ func (s *Server) listen(addr string) {
 }
 
 //
+// websocket listen goroutine
+//
+
+func (s *Server) wslisten(addr string) {
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
```
Comment (edmund-huber) [inline irc/server.go:216]: If we do allow multiple `wslisten`, then we might have to be careful about setting up this HTTP method handler multiple times. Maybe just pull it out of the method?

File: irc/websocket.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/gorilla/websocket"
+	"net"
+	"net/http"
+	"time"
+)
+
+var upgrader = websocket.Upgrader{
+	ReadBufferSize:  1024,
+	WriteBufferSize: 1024,
+	CheckOrigin:     func(r *http.Request) bool { return true },
```
Comment (edmund-huber) [inline irc/websocket.go:13]: I think this merits some sort of comment. Why is any origin okay? What is CheckOrigin exactly? :)

File: irc/websocket.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+type WSContainer struct {
```
Comment (edmund-huber) [inline irc/websocket.go:16]: Would it make more sense to call this type `WSConn` to be consistent with the interface it's implementing, http://golang.org/pkg/net/#Conn ?

File: irc/websocket.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+func (this WSContainer) Write(msg []byte) (int, error) {
+	err := this.conn.WriteMessage(1, msg)
+	return len(msg), err
+}
+
+func (this WSContainer) SetDeadline(t time.Time) error {
+	err := this.conn.SetWriteDeadline(t)
+	err = this.conn.SetReadDeadline(t)
```
Comment (edmund-huber) [inline irc/websocket.go:46]: Can you replace these two lines with just `return this.conn.SetDeadline(t)`? In any case, the second lines shadows the first `err`.

File: irc/websocket.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+func (this WSContainer) Read(msg []byte) (int, error) {
+	_, tmp, err := this.conn.ReadMessage()
+	str := (string)(tmp)
+	n := copy(msg, ([]byte)(str+CRLF+CRLF))
+	return n, err
+}
+
+func (this WSContainer) Write(msg []byte) (int, error) {
+	err := this.conn.WriteMessage(1, msg)
```
Comment (edmund-huber) [inline irc/websocket.go:40]: Can you replace the `1` with `websocket.BinaryMessage` or `websocket.TextMessage` as appropriate?

**Comment by edmund-huber** [comment]
OK, looking good, here's a test trace (ergonomadic output):

```
2015/05/04 22:53:05 127.0.0.1:60165 → PASS test
2015/05/04 22:53:15 127.0.0.1:60165 → NICK edmundinho
2015/05/04 22:53:23 127.0.0.1:60165 → USER edmundinho edmundinho localhost :Edmund Huber
2015/05/04 22:53:23 127.0.0.1:60165 ← :localhost 001 edmundinho :Welcome to the Internet Relay Network edmundinho!edmundinho@localhost
2015/05/04 22:53:23 127.0.0.1:60165 ← :localhost 002 edmundinho :Your host is localhost, running version ergonomadic-1.4.4
2015/05/04 22:53:23 127.0.0.1:60165 ← :localhost 004 edmundinho localhost ergonomadic-1.4.4 aio beIikntPpTl
2015/05/04 22:53:23 127.0.0.1:60165 ← :localhost 422 edmundinho :MOTD File is missing
2015/05/04 22:53:30 127.0.0.1:60165 → JOIN #test
2015/05/04 22:53:30 127.0.0.1:60165 ← :edmundinho!edmundinho@localhost JOIN #test
2015/05/04 22:53:30 127.0.0.1:60165 ← :localhost 353 edmundinho = #test :@edmundinho
2015/05/04 22:53:30 127.0.0.1:60165 ← :localhost 366 edmundinho #test :End of NAMES list
2015/05/04 22:54:30 127.0.0.1:60165 ← PING :localhost
```

I have to do the PONGs myself right?

Then please just address or comment on the things I pointed out and we're good to go. :)

File: ergonomadic.conf (hunk 1 above)
Comment (stumpyfr) [inline ergonomadic.conf:2]: nop, only test

File: irc/server.go (hunk 3 above)
Comment (stumpyfr) [inline irc/server.go:216]: not super easy to listen with multiple http handler on different port, I need to check if I can do that without too many code.
We can add that for in a new PR

File: irc/websocket.go (hunk 4 above)
Comment (stumpyfr) [inline irc/websocket.go:13]: If you dont accept from any origin, the server will not accept connection from client, by "conception" they are not on the same origin.
The js code (the client), is not hosted by the IRC server but by the webpage who wants to add a chat system.

File: irc/websocket.go (hunk 5 above)
Comment (stumpyfr) [inline irc/websocket.go:16]: Totally, I was not sure about that name and pushed before think to another name, my fault.

**Comment by stumpyfr** [comment]
Some inline comment, I will make some fix later, I am currently in travel, difficult to find time on my computer.

File: irc/server.go (hunk 3 above)
Comment (edmund-huber) [inline irc/server.go:216]: Yeah, I agree -- put off multiple WS listen.

**Comment by edmund-huber** [comment]
@stumpyfr ping, how long do you think fixups will take? Do you mind if I take a whack if you are busy? :)

**Comment by stumpyfr** [comment]
Hi,

back from my travel but a little bit on rush, if you have time dont hesitate to move forward.
I will not be able to do that before next weekend...in the best case.

**Comment by edmund-huber** [comment]
Continued in https://github.com/edmund-huber/ergonomadic/pull/14.

**Pull request closed without merging**